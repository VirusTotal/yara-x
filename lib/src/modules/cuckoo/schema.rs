use std::fmt;

use serde::de::Error;
use serde::{de::Visitor, Deserialize, Deserializer};

#[derive(serde::Deserialize, Debug)]
pub(super) struct DomainJson {
    pub domain: String,
}

#[derive(serde::Deserialize, Debug)]
pub(super) struct HttpJson {
    #[serde(rename = "user-agent")]
    pub user_agent: Option<String>,
    pub method: Option<String>, // string ftw
    pub uri: String,
}

#[derive(serde::Deserialize, Debug)]
pub(super) struct TcpJson {
    pub dst: Option<String>,
    pub dst_domain: Option<String>,
    pub dport: u64,
}

#[derive(serde::Deserialize, Debug)]
pub(super) struct UdpJson {
    pub dst: Option<String>,
    pub dst_domain: Option<String>,
    pub dport: u64,
}

#[derive(/* serde::Deserialize, - custom */ Debug)]
pub(super) struct NetworkJson {
    pub domains: Option<Vec<DomainJson>>,
    pub http: Option<Vec<HttpJson>>,
    pub tcp: Option<Vec<TcpJson>>,
    pub udp: Option<Vec<UdpJson>>,
    pub hosts: Option<Vec<String>>,
}

#[derive(serde::Deserialize, Debug)]
pub(super) struct SummaryJson {
    pub mutexes: Option<Vec<String>>,
    pub files: Option<Vec<String>>,
    pub keys: Option<Vec<String>>,
}

#[derive(serde::Deserialize, Debug)]
pub(super) struct BehaviorJson {
    pub summary: SummaryJson,
}

#[derive(serde::Deserialize, Debug)]
pub(super) struct CuckooJson {
    pub network: NetworkJson,
    pub behavior: BehaviorJson,
}

impl<'de> Deserialize<'de> for NetworkJson {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct MyVisitor;

        impl<'de> Visitor<'de> for MyVisitor {
            type Value = NetworkJson;

            fn expecting(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
                fmt.write_str("string or object")
            }

            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
            where
                A: serde::de::MapAccess<'de>,
            {
                // Must not parse `old_domains` before the whole map is
                // searched if there is a `domains` field, then the value for
                // the key `old_domains` should be ignored - specifically, it
                // is okay if the `old_domains` does not have the expected
                // structure if `domains` is present.
                let mut old_domains = None::<serde_json::Value>;
                let mut domains = None::<serde_json::Value>;

                let mut http = None::<Vec<HttpJson>>;
                let mut tcp = None::<Vec<TcpJson>>;
                let mut udp = None::<Vec<UdpJson>>;
                let mut hosts = None::<Vec<String>>;

                while let Some((key, val)) =
                    map.next_entry::<String, serde_json::Value>()?
                {
                    match key.as_str() {
                        "domains" => {
                            domains = Some(val);
                        }
                        "dns" => {
                            if domains.is_some() {
                                continue; // prefer "domains" over "dns"
                            }
                            old_domains = Some(val);
                        }
                        "http" => {
                            http = Some(
                                Deserialize::deserialize(val)
                                    .map_err(Error::custom)?,
                            );
                        }
                        "tcp" => {
                            tcp = Some(
                                Deserialize::deserialize(val)
                                    .map_err(Error::custom)?,
                            );
                        }
                        "udp" => {
                            udp = Some(
                                Deserialize::deserialize(val)
                                    .map_err(Error::custom)?,
                            );
                        }
                        "hosts" => {
                            hosts = Some(
                                Deserialize::deserialize(val)
                                    .map_err(Error::custom)?,
                            );
                        }
                        _ => {}
                    }
                }

                #[derive(serde::Deserialize, Debug)]
                struct OldDomainJson {
                    pub hostname: String,
                }

                let domains: Option<Vec<DomainJson>> =
                    match (domains, old_domains) {
                        (Some(domains), _) => {
                            Deserialize::deserialize(domains)
                                .map_err(Error::custom)?
                        }
                        (None, Some(old_domains)) => {
                            let old_domains: Vec<OldDomainJson> =
                                Deserialize::deserialize(old_domains)
                                    .map_err(Error::custom)?;

                            Some(
                                old_domains
                                    .into_iter()
                                    .map(|old| DomainJson {
                                        domain: old.hostname,
                                    })
                                    .collect(),
                            )
                        }
                        (None, None) => None, // domains field is optional
                    };

                Ok(NetworkJson { domains, http, tcp, udp, hosts })
            }
        }

        deserializer.deserialize_any(MyVisitor)
    }
}
