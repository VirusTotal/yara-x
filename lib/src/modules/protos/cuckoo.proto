syntax = "proto2";
import "yara.proto";

package cuckoo;

option (yara.module_options) = {
  name : "cuckoo"
  root_message: "cuckoo.Cuckoo"
  rust_module: "cuckoo"
  cargo_feature: "cuckoo-module"
};

message Cuckoo {
  // This module contains only exported functions, and doesn't return any data
}