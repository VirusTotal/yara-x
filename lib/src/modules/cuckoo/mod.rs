#[cfg(feature = "logging")]
use log::error;

use crate::compiler::RegexpId;
use crate::modules::prelude::*;
use crate::modules::protos::cuckoo::*;

mod schema;
#[cfg(test)]
mod tests;

use std::cell::RefCell;
use std::rc::Rc;
thread_local! {
    static LOCAL_DATA: RefCell<Option<Rc<schema::CuckooJson>>> = const { RefCell::new(None) };
}

fn get_local() -> Option<Rc<schema::CuckooJson>> {
    LOCAL_DATA.with(|data| data.borrow().clone())
}

fn set_local(value: schema::CuckooJson) {
    LOCAL_DATA.with(|data| {
        *data.borrow_mut() = Some(Rc::new(value));
    });
}

#[module_main]
fn main(_data: &[u8], meta: Option<&[u8]>) -> Cuckoo {
    let parsed =
        serde_json::from_slice::<schema::CuckooJson>(meta.unwrap_or_default());

    match parsed {
        Ok(parsed) => {
            set_local(parsed);
        }
        #[cfg(feature = "logging")]
        Err(e) => {
            error!("can't parse cuckoo report: {}", e);
        }
        #[cfg(not(feature = "logging"))]
        Err(_) => {}
    };

    Cuckoo::new()
}

#[module_export(name = "network.dns_lookup")]
fn network_dns_lookup_r(
    ctx: &ScanContext,
    regexp_id: RegexpId,
) -> Option<i64> {
    Some(
        get_local()?
            .network
            .domains
            .iter()
            .flatten()
            .filter(|domain| {
                ctx.regexp_matches(regexp_id, domain.domain.as_bytes())
            })
            .count() as _,
    )
}

#[module_export(name = "network.http_request")]
fn network_http_request_r(
    ctx: &ScanContext,
    regexp_id: RegexpId,
) -> Option<i64> {
    Some(
        get_local()?
            .network
            .http
            .iter()
            .flatten()
            .filter(|http| {
                http.method.is_some() // ~> is request (is not response)
                    && ctx.regexp_matches(regexp_id, http.uri.as_bytes())
            })
            .count() as _,
    )
}

#[module_export(name = "network.http_get")]
fn network_http_get_r(ctx: &ScanContext, regexp_id: RegexpId) -> Option<i64> {
    Some(
        get_local()?
            .network
            .http
            .iter()
            .flatten()
            .filter(|http| {
                http.method
                    .as_ref()
                    .map(|method| method.eq_ignore_ascii_case("get"))
                    .unwrap_or(false)
                    && ctx.regexp_matches(regexp_id, http.uri.as_bytes())
            })
            .count() as _,
    )
}

#[module_export(name = "network.http_post")]
fn network_http_post_r(ctx: &ScanContext, regexp_id: RegexpId) -> Option<i64> {
    Some(
        get_local()?
            .network
            .http
            .iter()
            .flatten()
            .filter(|http| {
                http.method
                    .as_ref()
                    .map(|method| method.eq_ignore_ascii_case("post"))
                    .unwrap_or(false)
                    && ctx.regexp_matches(regexp_id, http.uri.as_bytes())
            })
            .count() as _,
    )
}

#[module_export(name = "network.http_user_agent")]
fn network_http_user_agent_r(
    ctx: &ScanContext,
    regexp_id: RegexpId,
) -> Option<i64> {
    Some(
        get_local()?
            .network
            .http
            .iter()
            .flatten()
            .flat_map(|http| http.user_agent.iter())
            .filter(|user_agent| {
                ctx.regexp_matches(regexp_id, user_agent.as_bytes())
            })
            .count() as _,
    )
}

#[module_export(name = "network.tcp")]
fn network_tcp_ri(
    ctx: &ScanContext,
    dst_re: RegexpId,
    port: i64,
) -> Option<i64> {
    Some(
        get_local()?
            .network
            .tcp
            .iter()
            .flatten()
            .filter(|tcp| {
                tcp.dport == port as u64
                    && tcp
                        .dst
                        .iter()
                        .chain(tcp.dst_domain.iter())
                        .any(|dst| ctx.regexp_matches(dst_re, dst.as_bytes()))
            })
            .count() as _,
    )
}

#[module_export(name = "network.udp")]
fn network_udp_ri(
    ctx: &ScanContext,
    dst_re: RegexpId,
    port: i64,
) -> Option<i64> {
    Some(
        get_local()?
            .network
            .udp
            .iter()
            .flatten()
            .filter(|udp| {
                udp.dport == port as u64
                    && udp
                        .dst
                        .iter()
                        .chain(udp.dst_domain.iter())
                        .any(|dst| ctx.regexp_matches(dst_re, dst.as_bytes()))
            })
            .count() as _,
    )
}

#[module_export(name = "network.host")]
fn network_host_r(ctx: &ScanContext, re: RegexpId) -> Option<i64> {
    Some(
        get_local()?
            .network
            .hosts
            .iter()
            .flatten()
            .filter(|host| ctx.regexp_matches(re, host.as_bytes()))
            .count() as _,
    )
}

#[module_export(name = "sync.mutex")]
fn sync_mutex_r(ctx: &ScanContext, mutex_re: RegexpId) -> Option<i64> {
    Some(
        get_local()?
            .behavior
            .summary
            .mutexes
            .iter()
            .flatten()
            .filter(|mutex| ctx.regexp_matches(mutex_re, mutex.as_bytes()))
            .count() as _,
    )
}

#[module_export(name = "filesystem.file_access")]
fn filesystem_file_access_r(
    ctx: &ScanContext,
    regexp_id: RegexpId,
) -> Option<i64> {
    Some(
        get_local()?
            .behavior
            .summary
            .files
            .iter()
            .flatten()
            .filter(|file| ctx.regexp_matches(regexp_id, file.as_bytes()))
            .count() as _,
    )
}

#[module_export(name = "registry.key_access")]
fn registry_key_access_r(
    ctx: &ScanContext,
    regexp_id: RegexpId,
) -> Option<i64> {
    Some(
        get_local()?
            .behavior
            .summary
            .keys
            .iter()
            .flatten()
            .filter(|key| ctx.regexp_matches(regexp_id, key.as_bytes()))
            .count() as _,
    )
}
