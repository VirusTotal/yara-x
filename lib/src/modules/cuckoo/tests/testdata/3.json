{
  "info": {
    "started": "2018-11-12 19:43:44",
    "version": "v0.1"
  }
}