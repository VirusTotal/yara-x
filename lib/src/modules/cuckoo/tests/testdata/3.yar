import "cuckoo"

rule checkip {
	condition:
	  cuckoo.network.http_request(/.*/) == 0
}

rule checkip_get {
	condition:
	  not cuckoo.network.http_get(/.*/)
}

rule checkip_post {
	condition:
	  not cuckoo.network.http_post(/.*/)
}

rule checkip_dns {
	condition:
	  not cuckoo.network.dns_lookup(/.*/)
}

rule user_agent {
	condition:
	  not cuckoo.network.http_user_agent(/.*/)
}

rule mutex {
	condition:
	  not cuckoo.sync.mutex(/.*/)
}

rule file_access {
	condition:
	  not cuckoo.filesystem.file_access(/.*/)
}

rule reg {
	condition:
      cuckoo.registry.key_access(/.*/) == 0
}
