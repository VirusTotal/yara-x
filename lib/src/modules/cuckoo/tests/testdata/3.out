MATCHING RULES
--------------

checkip
checkip_get
checkip_post
checkip_dns
user_agent
mutex
file_access
reg

NON-MATCHING RULES
------------------

